Isolate Python SDK snippets from PR #6356

alnoki opened the pull request:
Per [request](https://github.com/aptos-labs/aptos-core/pull/6356#issuecomment-1445483042) of @davidiw in #6356 , isolates multisig Python SDK snippets from AMEE content.

The following commands were run in below paths:

aptos-core/:
```
pre-commit run --all-files
```

aptos-core/developer-docs-site/:
```
prettier '**/*.(tsx|ts|js)'
pnpm spellcheck
pnpm lint
```

aptos-core/ecosystem/python/sdk:
```
make fmt
make test
```

Modifications to `bcs.py` as a result of `make fmt` were disregarded since the file's contents were not modified per this PR.

davidiw commented on aptos-move/move-examples/upgrade_and_govern/v1_0_0/Move.toml:
```diff
@@ -0,0 +1,10 @@
+# :!:>manifest
+[package]
+name = 'UpgradeAndGovern'
+version = '1.0.0'
```
rather than versioning, could we just label these with more expressive names?

davidiw commented on aptos-move/move-examples/upgrade_and_govern/v1_1_0/scripts/set_and_transfer.move:
```diff
@@ -0,0 +1,18 @@
+// :!:>script
+script {
+    use upgrade_and_govern::parameters;
```
I would avoid placing multiple scripts in the same package and entirely from modules, reason being, they don't keep their names as they get compiled, so you don't know which script is which.

alnoki replied:
> they don't keep their names as they get compiled, so you don't know which script is which

@davidiw I'm a bit confused by this guidance, as the upgrade package has two scripts, `set_and_transfer.move` and `set_only.move`, which respectively compile into `set_and_transfer.mv` and `set_only.mv` as expected.

I suspect the hypothetical name loss you refer to is a result of using `main` as the main function name for every script in a Move package:

```move
script {
    fun main( // <- main function name
    //  ^ this PR uses either `set_only` or `set_and_transfer`, but
    //    name loss would result if both had `main`
        ...
    ) {
        ...
    }
}
```

This is analogous to using the same module definition block across multiple module `.move` files

```move
module named_address::a_repeated_module_name {
    //                ^ this compiles into `a_repeated_module_name.mv`
    //                  regardless of the `.move` filename
    ...
}
```

What are your thoughts?

davidiw commented on ecosystem/python/sdk/aptos_sdk/account.py:
```diff
@@ -79,6 +80,37 @@ def public_key(self) -> ed25519.PublicKey:
         return self.private_key.public_key()
 
 
+class RotationProofChallenge:
```
so I was suggesting that we emit out a chunk of bytes from say https://github.com/aptos-labs/aptos-core/blob/58214fc22ec00ebaeda467416e27574cdc6eb01d/aptos-move/e2e-move-tests/src/tests/rotate_auth_key.rs#L162

and dumping this to hex and importing that hex string here... that way we can be confident that it works as expected.

alnoki replied:
@davidiw how does this look?

https://github.com/alnoki/aptos-core/blob/dc8ac59fe89ecf04f2b6187a7e2ae83179cde23a/ecosystem/python/sdk/aptos_sdk/account.py#L131-L160

The classes have been verified e2e on devnet using the example script broken down in the tutorial, so I just emitted some bytes and made a test for it.

davidiw replied:
great

davidiw commented on ecosystem/python/sdk/aptos_sdk/ed25519.py:
```diff
@@ -92,6 +94,42 @@ def serialize(self, serializer: Serializer):
         serializer.to_bytes(self.key.encode())
 
 
+class MultiEd25519PublicKey:
```
probably should validate threshold and number of keys

maybe move to another file as well?

alnoki replied:
@davidiw requested range checks and expected failure tests have been added.

davidiw commented on ecosystem/python/sdk/aptos_sdk/ed25519.py:
```diff
@@ -92,6 +94,42 @@ def serialize(self, serializer: Serializer):
+        return concatenated_keys + bytes([self.threshold])
+
+    @staticmethod
+    def from_bytes(key: bytes) -> MultiEd25519PublicKey:
```
should validate threshold and length

alnoki replied:
@davidiw requested range checks and tests have been added.

davidiw commented on ecosystem/python/sdk/aptos_sdk/ed25519.py:
```diff
@@ -92,6 +94,42 @@ def serialize(self, serializer: Serializer):
+            keys.append(PublicKey(VerifyKey(key[start_byte:end_byte])))
+        return MultiEd25519PublicKey(keys, int(key[-1]))
+
+    def serialize(self, serializer: Serializer):
```
same here, we have no tests, we could probably generate some strings in the aptos rust code as samples

at least have some e2e tests that show that a set of public keys can verify a signature

alnoki replied:
@davidiw e2e tests added.

davidiw commented on ecosystem/python/sdk/aptos_sdk/ed25519.py:
```diff
@@ -123,6 +161,35 @@ def serialize(self, serializer: Serializer):
+            concatenated_signatures += signature.data()
+        return concatenated_signatures + self.bitmap
+
+    def serialize(self, serializer: Serializer):
```
would benefit from e2e tests and validation.

davidiw replied:
even if you can't add bytes for tests, we can still sign and verify

alnoki replied:
@davidiw See new tests.

davidiw commented on ecosystem/python/sdk/examples/multisig.py:
```diff
@@ -0,0 +1,483 @@
+from aptos_sdk.type_tag import StructTag, TypeTag
+from common import FAUCET_URL, NODE_URL
+
+wait_for_user = True
```
not sure this adds much value, can we remove it?

alnoki replied:
@davidiw removed.

davidiw commented on ecosystem/python/sdk/examples/multisig.py:
```diff
@@ -0,0 +1,483 @@
+
+    print(f"Auth key pre-rotation: {auth_key}")
+
+    rest_client.submit_bcs_transaction(signed_transaction)
+
+    print("\nWaiting for API server to update...\n")
+    time.sleep(2.5)
```
we have a wait for txn, why not use it?

alnoki replied:
@davidiw updated.

davidiw commented on ecosystem/python/sdk/examples/multisig.py:
```diff
@@ -0,0 +1,483 @@
+    packages_dir = "../../../../aptos-move/move-examples/upgrade_and_govern/"
+
+    command = (
+        f"aptos move compile --save-metadata "
```
maybe move --save-metadata to anotehr line and make one per line

alnoki replied:
@davidiw updated.

davidiw reviewed (commented):
Thanks! @clay-aptos want to do a quick run through of the documentation?

clay-aptos reviewed (approved):
Thanks so much, Al. The docs LGTM with some tiny changes I made in direct commits and a fix to the examples directory I flag in a comment here. Thanks again! I look forward to adding this to our Community Highlights list!

clay-aptos commented:
Hi @alnoki , I am noting we launched this:
https://aptos.dev/community/contributions/community-highlights

And would love to have this online. Please let me know how O can help. Thanks!

alnoki commented:
> Hi @alnoki , I am noting we launched this: https://aptos.dev/community/contributions/community-highlights
> 
> And would love to have this online. Please let me know how O can help. Thanks!

@clay-aptos thank you for your patience during all of the delays during Aptos Week at Mountain DAO and then ETH Denver. I've about 80% done with a commit that addresses comments from you and @davidiw above and am hoping to push in the next few days or so.

alnoki commented on developer-docs-site/docs/tutorials/first-multisig.md:
```diff
@@ -0,0 +1,359 @@
+
+## Step 8: Perform Move package governance
+
+In this section, the multisig vanity account will publish a simple package, upgrade it, then invoke a [Move governance](https://github.com/aptos-labs/aptos-core/tree/main/aptos-move/move-examples/upgrade_and_govern) script.
```
@clay-aptos the link is one that would be valid upon landing of this PR (so not yet in the `aptos-core` repo, and I've updated the sentence in question for clarity. How does it look now?

alnoki replied:
Resolving per https://github.com/aptos-labs/aptos-core/pull/6788#issuecomment-1472877448

alnoki commented:
@davidiw @clay-aptos I believe I've addressed your review comments above: is there anything else you need to move this along?

clay-aptos commented:
> @davidiw @clay-aptos I believe I've addressed your review comments above: is there anything else you need to move this along?

Hi @alnoki , this LGTM! Thank you.  @davidiw ?

davidiw commented on ecosystem/python/sdk/examples/multisig.py:
```diff
@@ -0,0 +1,468 @@
+    print(f"Bob:   {bob.public_key()}")
+    print(f"Chad:  {chad.public_key()}")  # <:!:section_1
+
+    wait()
```
what's the point of these waits? we want these to run in a test environment... we need to quickly follow up this PR with something that validates this on every PR -- these quickly bitrot or start failing due to other ninjas

davidiw reviewed (approved):
thanks!!

davidiw commented:
realized the tutorial is interactive... this will bitrot, we need to consider some other means, even if that is that we keep pre-compiled versions around.

clay-aptos commented:
Just noting we are announcing and linking in:
https://github.com/aptos-labs/aptos-core/pull/7308#issuecomment-1478170772
